### PR TITLE
[ADD] add module to manage product by business

### PR DESCRIPTION
**Pull request by laetitia-gangloff**
First part : add new model (business.product.location) to group some product/quantity into a business and attach these business to a location.

Second part : a report with product/quantity not in the location but excepted and product/quantity not excepted in the location.

### CODE REVIEW

File: business_product_location/README.rst [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+Business product location
+===========
+
+The goal of this module is to manage a list of product and quantity attached
+to a business and then define a business usage on stock location.
+To begin define a list of product that must be in a stock location.
+Depending on the business usage defined on the location, the list of product
+is different, Locations with the same business usage always have the same
+product list. He also need to compare the theoretical list  with the real
```
Comment (ghost) [inline business_product_location/README.rst]: who is 'he' ?

File: business_product_location/models/business_product_location.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                        product_line.product_id.id] = {
+                        'product_name': product_line.product_id.display_name,
+                        'excepted_qty': product_line.product_qty,
+                        'real_qty': 0,
+                        'outgoing_qty': 0,
+                        'strong': True}
+                # get product in this location
+                product_in_loc = product_obj.with_context(location=location.id
+                                                          ).search(
+                    [('qty_available', '>', 0)])
```
Comment (ghost) [inline business_product_location/models/business_product_location.py]: This syntax is very strange. Please use 2 lines.
product_in_loc = product_obj.with_context(
    location=location.id).search([('qty_available', '>', 0)])

File: business_product_location/models/business_product_location.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+        less_product_by_loc = {}
+        more_product_by_loc = {}
+
+        for location in docs:
+            if location.business_usage_id:
+                location_name = '%s - %s' % (location.name,
+                                             location.business_usage_id.name)
+                more_product_by_loc[location_name] = {}
+                less_product_by_loc[location_name] = {}
```
Comment (ghost) [inline business_product_location/models/business_product_location.py]: Using a string as dictionary key is most time a bad idea. Nothing makes you sure that you won't somehow have 2 identical tuples of (location.name, location.business_usage_id.name). Also, using a tuple of ids (location.id, location.business_usage_id.id) would be more efficient.

File: business_product_location/models/business_product_location.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+                            product_to_update['outgoing_qty'],
+                            precision_rounding=1) == 0:
+                        product_to_update['strong'] = False
+                    if product.id not in less_product_by_loc[location_name
+                                                             ].keys():
+                        more_product_by_loc[location_name][
+                            product.id] = product_to_update
+                    else:
+                        less_product_by_loc[location_name][
+                            product.id] = product_to_update
```
Comment (ghost) [inline business_product_location/models/business_product_location.py]: Consider split your method (render_html) into different functions. The indentation makes it hard to read.

**Comment by laetitia-gangloff** [comment]
@dufresnedavid thank you for your suggestions !

I splitted render_html and tried to add more comments. For the key, I thought location name was unique, but it is not the case, so I updated.

**Comment by ghost** [comment]
@laetitia-gangloff Thanks :+1:

File: business_product_location/__openerp__.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+{
+    "name": "Business Product Location",
+    "version": "0.1",
+    'author': "Acsone, Odoo Community Association (OCA)",
```
Comment (lmignon) [inline business_product_location/__openerp__.py]: ACSONE SA/NV

File: business_product_location/__openerp__.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    "category": "Other",
```
Comment (lmignon) [inline business_product_location/__openerp__.py]: 'category': 'Warehouse Management',

File: business_product_location/models/business_product_location.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+
+class StockLocation(models.Model):
+    _inherit = 'stock.location'
+
+    business_usage_id = fields.Many2one('business.product.location',
+                                        'Business Usage')
+
+
+class BusinessProductLocationReport(models.AbstractModel):
```
Comment (lmignon) [inline business_product_location/models/business_product_location.py]: According to the OCA guidelines, reports should be in a report module. https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#directories

File: business_product_location/tests/test_business_product_location.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+import openerp.tests.common as common
+
+
+class TestBusinessProductLocation(common.TransactionCase):
+
+    def setUp(self):
```
Comment (lmignon) [inline business_product_location/tests/test_business_product_location.py]: This override is not needed.

File: business_product_location/tests/test_business_product_location.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+                * product_product_35 qty 3
+                * stock_location_stock
+            Check product_product_34 as a business_product_location with bpl
+            Check product_product_35 as a business_product_location with bpl
+            Check stock_location_stock as a business_product_location with bpl
+        """
+        bpl = self.env['business.product.location'].create(
+            {'name': 'Test BPL',
+             'product_ids': [
+                 (0, 0, {'product_id': self.env.ref(
```
Comment (lmignon) [inline business_product_location/tests/test_business_product_location.py]: PLZ assign self.env.ref('product.product_product_3X) to a variable to avoir copy/paste and improve the readability

File: business_product_location/tests/test_business_product_location.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+                     'product.product_product_34').id, 'product_qty': 4}),
+                 (0, 0, {'product_id': self.env.ref(
+                     'product.product_product_35').id, 'product_qty': 3})],
+             'location_ids': [(4,
+                               self.env.ref('stock.stock_location_stock').id)]
```
Comment (lmignon) [inline business_product_location/tests/test_business_product_location.py]: Idem

File: business_product_location/views/business_product_location_templates.xml [hunk 11]
```diff
@@ -0,0 +1,69 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+    <data>
+        <template id="report_bpl">
+            <t t-call="report.html_container">
+                <t t-foreach="less_product_by_loc" t-as="o">
+                    <t t-call="report.external_layout">
+                        <div class="page">
+                            <h2>
+                            Excepted product - <span t-esc="o_value['location_name']"/>
```
Comment (lmignon) [inline business_product_location/views/business_product_location_templates.xml:10]: Is this label translatable?

**Comment by lmignon** [comment]
The functionality is interesting. A better description would be welcome.

File: business_product_location/views/business_product_location_templates.xml (hunk 11 above)
Comment (laetitia-gangloff) [inline business_product_location/views/business_product_location_templates.xml:10]: It uses the lang of the current user.

**Comment by gurneyalex** [comment]
can you please rebase your PR? The travis build of 8.0 has recently been fixed.

**Comment by laetitia-gangloff** [comment]
It is rebased. Thank you.

**Comment by gurneyalex** [comment]
:+1:

**Comment by Cedric-Pigeon** [comment]
+1

**Review by Cedric-Pigeon — APPROVED** [review]

**Comment by Cedric-Pigeon** [comment]
@pedrobaeza I think this one can be merged

**Review by lmignon — APPROVED** [review]
:+1: Works in production for a long time

**Comment by pedrobaeza** [comment]
README file is outdated, but I won't be picky.

**Pull request merged**